File: packages/fortress/ios/Sources/FortressPlugin/Impl/StandardStorage.swift
import Foundation

/**
 Standard (insecure) storage implementation using UserDefaults.
 *
 * This layer provides non-encrypted storage for non-sensitive data
 * using iOS UserDefaults.
 *
 * Use for:
 * - UI theme preferences
 * - Non-sensitive flags
 * - Public configuration
 *
 * DO NOT use for:
 * - Passwords
 * - Authentication tokens
 * - Any sensitive data
 *
 * Architectural rules:
 * - Pure Swift implementation
 * - Stateless - no internal state
 * - Delegates to UserDefaults for persistence
 */
struct StandardStorage {

    // MARK: - Constants

    private static let suiteName = "io.capkit.fortress.standard"

    // MARK: - Public API

    /**
     Stores a value in standard (insecure) storage.

     - Parameters:
     - key: The unique key identifier.
     - value: The string value to store.
     - Throws: NativeError if storage fails.
     */
    func set(key: String, value: String) throws {
        let defaults = UserDefaults.standard
        defaults.set(value, forKey: key)
    }

    /**
     Retrieves a value from standard (insecure) storage.

     - Parameter key: The unique key identifier.
     - Returns: The stored string value, or nil if not found.
     */
    func get(key: String) -> String? {
        let defaults = UserDefaults.standard
        return defaults.string(forKey: key)
    }

    /**
     Removes a value from standard (insecure) storage.

     - Parameter key: The unique key identifier.
     */
    func remove(key: String) throws {
        let defaults = UserDefaults.standard
        defaults.removeObject(forKey: key)
    }

    /**
     Clears all values from standard storage.

     Note: This only clears keys that start with the fortress prefix.
     */
    func clearAll() throws {
        let defaults = UserDefaults.standard
        let allKeys = defaults.dictionaryRepresentation().keys

        for key in allKeys where key.hasPrefix("ftrss_") || key.hasPrefix("fortress_") {
            defaults.removeObject(forKey: key)
        }
    }

    /**
     Checks whether a key exists in standard storage.

     - Parameter key: The unique key identifier.
     - Returns: true if the key exists, false otherwise.
     */
    func hasKey(key: String) -> Bool {
        let defaults = UserDefaults.standard
        return defaults.object(forKey: key) != nil
    }
}

File: packages/fortress/android/src/main/java/io/capkit/fortress/impl/StandardStorage.kt
package io.capkit.fortress.impl

import android.content.Context
import android.content.SharedPreferences

/**
 * Standard (insecure) storage implementation using SharedPreferences.
 *
 * This layer provides non-encrypted storage for non-sensitive data
 * using Android SharedPreferences.
 *
 * Use for:
 * - UI theme preferences
 * - Non-sensitive flags
 * - Public configuration
 *
 * DO NOT use for:
 * - Passwords
 * - Authentication tokens
 * - Any sensitive data
 *
 * Architectural rules:
 * - Pure Kotlin implementation
 * - Stateless - no internal state
 * - Delegates to SharedPreferences for persistence
 */
class StandardStorage(
  context: Context,
) {
  companion object {
    private const val SUITE_NAME = "io.capkit.fortress.standard"
  }

  private val prefs: SharedPreferences =
    context.getSharedPreferences(SUITE_NAME, Context.MODE_PRIVATE)

  /**
   * Stores a value in standard (insecure) storage.
   *
   * @param key Unique key identifier
   * @param value String value to store
   */
  fun set(
    key: String,
    value: String,
  ) {
    prefs.edit().putString(key, value).apply()
  }

  /**
   * Retrieves a value from standard (insecure) storage.
   *
   * @param key Unique key identifier
   * @return Stored string value, or null if not found
   */
  fun get(key: String): String? = prefs.getString(key, null)

  /**
   * Removes a value from standard (insecure) storage.
   *
   * @param key Unique key identifier
   */
  fun remove(key: String) {
    prefs.edit().remove(key).apply()
  }

  /**
   * Clears all values from standard storage.
   *
   * Note: This only clears keys that start with the fortress prefix.
   */
  fun clearAll() {
    val keysToRemove =
      prefs.all.keys.filter {
        it.startsWith("ftrss_") || it.startsWith("fortress_")
      }

    if (keysToRemove.isEmpty()) return

    prefs
      .edit()
      .apply {
        keysToRemove.forEach { key ->
          // Logical improvement: "Wipe-before-delete" to overwrite
          // data in RAM before removing the disk pointer
          putString(key, "DELETED")
          remove(key)
        }
      }.apply()
  }

  /**
   * Checks whether a key exists in standard storage.
   *
   * @param key Unique key identifier
   * @return true if the key exists, false otherwise
   */
  fun hasKey(key: String): Boolean = prefs.contains(key)
}
